<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="`SPOTlight`provides a method to deconvolute spatial transcriptomics
    spots using a seeded NMF approach along with visualization tools to assess
    the results. Spatially resolved gene expression profiles are key to
    understand tissue organization and function. However, novel spatial 
    transcriptomics (ST) profiling techniques lack single-cell resolution and 
    require a combination with single-cell RNA sequencing (scRNA-seq) 
    information to deconvolute the spatially indexed datasets. Leveraging the 
    strengths of both data types, we developed SPOTlight, a computational tool 
    that enables the integration of ST with scRNA-seq data to infer the location
    of cell types and states within a complex tissue. SPOTlight is centered 
    around a seeded non-negative matrix factorization (NMF) regression, 
    initialized using cell-type marker genes and non-negative least squares
    (NNLS) to subsequently deconvolute ST capture locations (spots).">
<title>`SPOTlight`: Spatial Transcriptomics Deconvolution • SPOTlight</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="`SPOTlight`: Spatial Transcriptomics Deconvolution">
<meta property="og:description" content="`SPOTlight`provides a method to deconvolute spatial transcriptomics
    spots using a seeded NMF approach along with visualization tools to assess
    the results. Spatially resolved gene expression profiles are key to
    understand tissue organization and function. However, novel spatial 
    transcriptomics (ST) profiling techniques lack single-cell resolution and 
    require a combination with single-cell RNA sequencing (scRNA-seq) 
    information to deconvolute the spatially indexed datasets. Leveraging the 
    strengths of both data types, we developed SPOTlight, a computational tool 
    that enables the integration of ST with scRNA-seq data to infer the location
    of cell types and states within a complex tissue. SPOTlight is centered 
    around a seeded non-negative matrix factorization (NMF) regression, 
    initialized using cell-type marker genes and non-negative least squares
    (NNLS) to subsequently deconvolute ST capture locations (spots).">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">SPOTlight</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/SPOTlight_kidney.html">Spatial Transcriptomics Deconvolution with `SPOTlight`</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MarcElosua/SPOTlight/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="welcome-to-spotlight-">Welcome to <code>SPOTlight</code> <img src="inst/extdata/SPOTlight.png" width="200" align="right"><a class="anchor" aria-label="anchor" href="#welcome-to-spotlight-"></a>
</h1></div>
<p><code>SPOTlight</code> provides a tool that enables the deconvolution of mixtures of cells from a single-cell reference. Originally developed for 10X’s Visium - spatial transcriptomics- technology, it can be used for all technologies that output mixtures of cells. It is compatible with Bioconductor’s <code>SingleCellExperiment</code> and <code>SpatialExperiment</code> classes as well as with <code>Seurat</code> objects. Furthermore, the package also provides visualization tools to assess the results of the deconvolution. Briefly, <code>SPOTlight</code> is based on finding topic profile signatures, by means of an NMFreg model, for each cell type and then optimizing the cell types proportions to fit the mixture we want to deconvolute.</p>
<p><img src="inst/extdata/schematic.png" width="600"></p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"SPOTlight"</span><span class="op">)</span>
<span class="co"># Or the devel version</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"SPOTlight"</span>, version <span class="op">=</span> <span class="st">"devel"</span><span class="op">)</span></code></pre></div>
<p>Alternatively, you can install it from GitHub using the <a href="https://github.com/hadley/devtools" class="external-link">devtools</a> package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span>
<span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"https://github.com/MarcElosua/SPOTlight"</span><span class="op">)</span></code></pre></div>
<div class="section level3">
<h3 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h3>
<ul>
<li>Elosua-Bayes M, Nieto P, Mereu E, Gut I, Heyn H (2021): <em>SPOTlight: seeded NMF regression to deconvolute spatial transcriptomics spots with single-cell transcriptomes</em>. <strong>Nucleic Acids Res</strong> 49(9):e50. <a href="doi:10.1093/nar/gkab043" class="uri">doi:10.1093/nar/gkab043</a>.</li>
</ul>
<hr>
</div>
<div class="section level3">
<h3 id="contributors">Contributors<a class="anchor" aria-label="anchor" href="#contributors"></a>
</h3>
<p>SPOTlight was originally developed by <a href="https://github.com/MarcElosua/" class="external-link">Marc Elosua Bayes</a> and has received substantial additional contributions from <a href="https://github.com/HelenaLC" class="external-link">Helena L. Crowell</a>.</p>
</div>
<div class="section level3">
<h3 id="issues---ideas">Issues - Ideas?<a class="anchor" aria-label="anchor" href="#issues---ideas"></a>
</h3>
<p><code>SPOTlight</code> is still under active development. We greatly welcome (and highly encourage!) all feedback, bug reports and suggestions for improvement <a href="https://github.com/MarcElosua/SPOTlight/issues" class="external-link">here</a>. <strong>Please make sure to raise issues with a <a href="https://www.tidyverse.org/help/" class="external-link">reproducible example</a> and the output of your <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code>.</strong></p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/MarcElosua/SPOTlight/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/MarcElosua/SPOTlight/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing SPOTlight</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Marc Elosua-Bayes <br><small class="roles"> Author, maintainer </small>  </li>
<li>Helena L. Crowell <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Marc Elosua-Bayes, Helena L. Crowell.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
