<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This is the backbone function which takes in single cell
  expression data to deconvolute spatial transcriptomics spots."><title>Deconvolution of mixture using single-cell data — SPOTlight • SPOTlight</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Deconvolution of mixture using single-cell data — SPOTlight"><meta property="og:description" content="This is the backbone function which takes in single cell
  expression data to deconvolute spatial transcriptomics spots."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SPOTlight</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/SPOTlight_kidney.html">Spatial Transcriptomics Deconvolution with `SPOTlight`</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MarcElosua/SPOTlight/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Deconvolution of mixture using single-cell data</h1>
      <small class="dont-index">Source: <a href="https://github.com/MarcElosua/SPOTlight/blob/HEAD/R/SPOTlight.R" class="external-link"><code>R/SPOTlight.R</code></a></small>
      <div class="d-none name"><code>SPOTlight.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This is the backbone function which takes in single cell
  expression data to deconvolute spatial transcriptomics spots.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="co"># S4 method for SingleCellExperiment,ANY</span>
<span class="fu">SPOTlight</span><span class="op">(</span>
  <span class="va">x</span>,
  <span class="va">y</span>,
  <span class="va">...</span>,
  assay <span class="op">=</span> <span class="st">"counts"</span>,
  groups <span class="op">=</span> <span class="fu">colLabels</span><span class="op">(</span><span class="va">x</span>, onAbsence <span class="op">=</span> <span class="st">"error"</span><span class="op">)</span>
<span class="op">)</span>

<span class="co"># S4 method for ANY,SingleCellExperiment</span>
<span class="fu">SPOTlight</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">...</span>, assay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span>

<span class="co"># S4 method for Seurat,ANY</span>
<span class="fu">SPOTlight</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">...</span>, slot <span class="op">=</span> <span class="st">"counts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, groups <span class="op">=</span> <span class="fu">Idents</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span>

<span class="co"># S4 method for ANY,Seurat</span>
<span class="fu">SPOTlight</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">...</span>, slot <span class="op">=</span> <span class="st">"counts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span>

<span class="co"># S4 method for ANY,dgCMatrix</span>
<span class="fu">SPOTlight</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">...</span>, slot <span class="op">=</span> <span class="st">"counts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span>

<span class="co"># S4 method for dgCMatrix,ANY</span>
<span class="fu">SPOTlight</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">...</span>, slot <span class="op">=</span> <span class="st">"counts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span>

<span class="co"># S4 method for ANY,DelayedMatrix</span>
<span class="fu">SPOTlight</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">...</span>, slot <span class="op">=</span> <span class="st">"counts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span>

<span class="co"># S4 method for DelayedMatrix,ANY</span>
<span class="fu">SPOTlight</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">...</span>, slot <span class="op">=</span> <span class="st">"counts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span>

<span class="co"># S4 method for ANY,ANY</span>
<span class="fu">SPOTlight</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">...</span><span class="op">)</span>

<span class="co"># S4 method for matrix,matrix</span>
<span class="fu">SPOTlight</span><span class="op">(</span>
  <span class="va">x</span>,
  <span class="va">y</span>,
  <span class="va">groups</span>,
  <span class="va">mgs</span>,
  n_top <span class="op">=</span> <span class="cn">NULL</span>,
  gene_id <span class="op">=</span> <span class="st">"gene"</span>,
  group_id <span class="op">=</span> <span class="st">"cluster"</span>,
  weight_id <span class="op">=</span> <span class="st">"weight"</span>,
  hvg <span class="op">=</span> <span class="cn">NULL</span>,
  scale <span class="op">=</span> <span class="cn">TRUE</span>,
  model <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ns"</span>, <span class="st">"std"</span><span class="op">)</span>,
  min_prop <span class="op">=</span> <span class="fl">0.01</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x, y</dt>
<dd><p>single-cell and mixture dataset, respectively. Can be a
numeric matrix, <code>SingleCellExperiment</code> or <code>SeuratObjecy</code>.</p></dd>
<dt>...</dt>
<dd><p>additional parameters.</p></dd>
<dt>assay</dt>
<dd><p>if the object is of Class <code>Seurat</code>, character string
specifying the assay from which to extract the expression matrix.
By default "RNA".</p></dd>
<dt>groups</dt>
<dd><p>vector of group labels for cells in <code>x</code>.
When <code>x</code> is a <code>SingleCellExperiment</code> or <code>SeuratObject</code>,
defaults to <code>colLabels</code> and <code>Idents(x)</code>, respectively.</p></dd>
<dt>slot</dt>
<dd><p>if the object is of Class <code>Seurat</code>, character string
specifying the slot from which to extract the expression matrix.
By default "counts".</p></dd>
<dt>mgs</dt>
<dd><p><code>data.frame</code> or <code>DataFrame</code> of marker genes.
Must contain columns holding gene identifiers, group labels and
the weight (e.g., logFC, -log(p-value) a feature has in a given group.</p></dd>
<dt>n_top</dt>
<dd><p>integer scalar specifying the number of markers to select per
group. By default NULL uses all the marker genes to initialize the model.</p></dd>
<dt>gene_id, group_id, weight_id</dt>
<dd><p>character specifying the column
in <code>mgs</code> containing gene identifiers, group labels and weights,
respectively.</p></dd>
<dt>hvg</dt>
<dd><p>character vector containing hvg to include in the model.
By default NULL.</p></dd>
<dt>scale</dt>
<dd><p>logical specifying whether to scale single-cell counts to unit
variance. This gives the user the option to normalize the data beforehand
as you see fit (CPM, FPKM, ...) when passing a matrix or specifying the
slot from where to extract the count data.</p></dd>
<dt>model</dt>
<dd><p>character string indicating which model to use when running NMF.
Either "ns" (default) or "std".</p></dd>
<dt>min_prop</dt>
<dd><p>scalar in [0,1] setting the minimum contribution
expected from a cell type in <code>x</code> to observations in <code>y</code>.
By default 0.</p></dd>
<dt>verbose</dt>
<dd><p>logical. Should information on progress be reported?</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a numeric matrix with rows corresponding to samples
  and columns to groups</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>SPOTlight uses a Non-Negative Matrix Factorization approach to learn
  which genes are important for each cell type. In order to drive the
  factorization and give more importance to cell type marker genes we
  previously compute them and use them to initialize the basis matrix. This
  initialized matrices will then be used to carry out the factorization with
  the single cell expression data. Once the model has learn the topic
  profiles for each cell type we use non-negative least squares (NNLS) to
  obtain the topic contributions to each spot. Lastly, NNLS is again used to
  obtain the proportion of each cell type for each spot by finding the
  fitting the single-cell topic profiles to the spots topic contributions.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Marc Elosua-Bayes &amp; Helena L. Crowell</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SingleCellExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SummarizedExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: MatrixGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: matrixStats</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘matrixStats’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:Biobase’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     anyMissing, rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘MatrixGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedSds, rowWeightedVars</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:Biobase’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomicRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: stats4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: S4Vectors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘S4Vectors’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     I, expand.grid, unname</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: IRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomeInfoDb</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: scuttle</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: ggplot2</span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: AnnotationHub</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: BiocFileCache</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: dbplyr</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘AnnotationHub’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:Biobase’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     cache</span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/helenalc/TENxVisiumData" class="external-link">TENxVisiumData</a></span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SpatialExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> snapshotDate(): 2021-10-19</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Initialize a Hub instance storing a complete set of records</span></span>
<span class="r-in"><span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> snapshotDate(): 2021-10-19</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Retrieve any records that match our keyword(s) of interest</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">eh</span>, <span class="st">"Tabula Muris Senis droplet Kidney"</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ExperimentHub with 7 records</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> # snapshotDate(): 2021-10-19</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> # $dataprovider: The Tabula Muris Consortium</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> # $species: Mus musculus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> # $rdataclass: matrix, H5File, DFrame</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> # additional mcols(): taxonomyid, genome, description,</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> #   rdatapath, sourceurl, sourcetype </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> # retrieve records with, e.g., 'object[["EH6386"]]' </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            title                                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   EH6386 | Tabula Muris Senis droplet Kidney colData         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   EH6387 | Tabula Muris Senis droplet Kidney counts          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   EH6388 | Tabula Muris Senis droplet Kidney processed counts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   EH6389 | Tabula Muris Senis droplet Kidney rowData         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   EH6390 | Tabula Muris Senis droplet Kidney PCA             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   EH6391 | Tabula Muris Senis droplet Kidney tSNE            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   EH6392 | Tabula Muris Senis droplet Kidney UMAP            </span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">eh</span>, <span class="st">"MouseKidneyCoronal"</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ExperimentHub with 2 records</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> # snapshotDate(): 2021-10-19</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> # $dataprovider: 10X Genomics</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> # $species: Mus musculus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> # $rdataclass: SpatialExperiment</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> # additional mcols(): taxonomyid, genome, description,</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> #   rdatapath, sourceurl, sourcetype </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> # retrieve records with, e.g., 'object[["EH6707"]]' </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            title                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   EH6707 | MouseKidneyCoronal      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   EH6743 | MouseKidneyCoronal_v3.13</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Get Visium data from 'TENxVisiumData'</span></span>
<span class="r-in"><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TENxVisiumData/man/TENxVisiumData.html" class="external-link">MouseKidneyCoronal</a></span><span class="op">(</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?TENxVisiumData and browseVignettes('TENxVisiumData') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Use symbols instead of Ensembl IDs as feature names</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">symbol</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Load SCE data</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/fmicompbio/TabulaMurisSenisData" class="external-link">TabulaMurisSenisData</a></span><span class="op">)</span></span>
<span class="r-in"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TabulaMurisSenisData/man/TabulaMurisSenisDroplet.html" class="external-link">TabulaMurisSenisDroplet</a></span><span class="op">(</span>tissues <span class="op">=</span> <span class="st">"Kidney"</span><span class="op">)</span><span class="op">$</span><span class="va">Kidney</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> snapshotDate(): 2021-10-19</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?TabulaMurisSenisData and browseVignettes('TabulaMurisSenisData') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> require(“rhdf5”)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?TabulaMurisSenisData and browseVignettes('TabulaMurisSenisData') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?TabulaMurisSenisData and browseVignettes('TabulaMurisSenisData') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?TabulaMurisSenisData and browseVignettes('TabulaMurisSenisData') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?TabulaMurisSenisData and browseVignettes('TabulaMurisSenisData') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?TabulaMurisSenisData and browseVignettes('TabulaMurisSenisData') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Keep cells from 18m mice with clear cell type annotations</span></span>
<span class="r-in"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">sce</span>, , <span class="va">age</span> <span class="op">==</span> <span class="st">"18m"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">sce</span>, , <span class="op">!</span> <span class="va">free_annotation</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nan"</span>, <span class="st">"CD45"</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Get the top 3000 highly variable genes</span></span>
<span class="r-in"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span class="r-in"><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span class="r-in"><span class="va">hvg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n <span class="op">=</span> <span class="fl">3000</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-colData.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">free_annotation</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Get vector indicating which genes</span></span>
<span class="r-in"><span class="co"># are neither ribosomal or mitochondrial</span></span>
<span class="r-in"><span class="va">genes</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^Rp[l|s]|Mt"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Compute marker genes</span></span>
<span class="r-in"><span class="va">mgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/scoreMarkers.html" class="external-link">scoreMarkers</a></span><span class="op">(</span><span class="va">sce</span>, subset.row <span class="op">=</span> <span class="va">genes</span><span class="op">)</span></span>
<span class="r-in"><span class="va">mgs_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mgs</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span class="r-in">  <span class="va">x</span> <span class="op">&lt;-</span> <span class="va">mgs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span></span>
<span class="r-in">  <span class="co"># Filter and keep relevant marker genes, those with AUC &gt; 0.8</span></span>
<span class="r-in">  <span class="va">x</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span><span class="va">x</span><span class="op">$</span><span class="va">mean.AUC</span> <span class="op">&gt;</span> <span class="fl">0.8</span>, <span class="op">]</span></span>
<span class="r-in">  <span class="co"># Sort the genes from highest to lowest weight</span></span>
<span class="r-in">  <span class="va">x</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">mean.AUC</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span></span>
<span class="r-in">  <span class="co"># Add gene and cluster id to the dataframe</span></span>
<span class="r-in">  <span class="va">x</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span class="r-in">  <span class="va">x</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="va">i</span></span>
<span class="r-in">  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span><span class="op">)</span></span>
<span class="r-in"><span class="va">mgs_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">mgs_ls</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># split cell indices by identity</span></span>
<span class="r-in"><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>, <span class="va">sce</span><span class="op">$</span><span class="va">free_annotation</span><span class="op">)</span></span>
<span class="r-in"><span class="co"># downsample to at most 20 cells per identity</span></span>
<span class="r-in"><span class="va">n_cells</span> <span class="op">&lt;-</span> <span class="fl">20</span></span>
<span class="r-in"><span class="va">cs_keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">idx</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in">  <span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span></span>
<span class="r-in">  <span class="kw">if</span> <span class="op">(</span><span class="va">n</span> <span class="op">&lt;</span> <span class="va">n_cells</span><span class="op">)</span></span>
<span class="r-in">    <span class="va">n_cells</span> <span class="op">&lt;-</span> <span class="va">n</span></span>
<span class="r-in">  <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">i</span>, <span class="va">n_cells</span><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span><span class="op">)</span></span>
<span class="r-in"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">cs_keep</span><span class="op">)</span><span class="op">]</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">SPOTlight</span><span class="op">(</span></span>
<span class="r-in">    x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>,</span>
<span class="r-in">    y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span>,</span>
<span class="r-in">    groups <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">free_annotation</span>,</span>
<span class="r-in">    mgs <span class="op">=</span> <span class="va">mgs_df</span>,</span>
<span class="r-in">    hvg <span class="op">=</span> <span class="va">hvg</span>,</span>
<span class="r-in">    weight_id <span class="op">=</span> <span class="st">"mean.AUC"</span>,</span>
<span class="r-in">    group_id <span class="op">=</span> <span class="st">"cluster"</span>,</span>
<span class="r-in">    gene_id <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Scaling count matrix</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Seeding initial matrices</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Training NMF model</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Time for training: 8.95min</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Deconvoluting mixture data</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Marc Elosua-Bayes, Helena L. Crowell.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer></div>

  

  

  </body></html>

