<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SPOTlight">
<title>Spatial Transcriptomics Deconvolution with `SPOTlight` • SPOTlight</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Spatial Transcriptomics Deconvolution with `SPOTlight`">
<meta property="og:description" content="SPOTlight">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SPOTlight</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/SPOTlight_kidney.html">Spatial Transcriptomics Deconvolution with `SPOTlight`</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MarcElosua/SPOTlight/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Spatial Transcriptomics Deconvolution with `SPOTlight`</h1>
                        <h4 data-toc-skip class="author">Marc Elosua-Bayes</h4>
            <address class="author_afil">
      National Center for Genomic Analysis - Center for Genomic RegulationUniversity Pompeu Fabra<br><a class="author_email" href="mailto:#"></a><a href="mailto:marc.elosua@cnag.crg.eu" class="email">marc.elosua@cnag.crg.eu</a>
      </address>
                              <h4 data-toc-skip class="author">Helena L. Crowell</h4>
            <address class="author_afil">
      Institute for Molecular Life Sciences, University of Zurich, SwitzerlandSIB Swiss Institute of Bioinformatics, University of Zurich, Switzerland<br><a class="author_email" href="mailto:#"></a><a href="mailto:helena.crowell@uzh.ch" class="email">helena.crowell@uzh.ch</a>
      </address>
                  
            <h4 data-toc-skip class="date">18 enero 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/MarcElosua/SPOTlight/blob/HEAD/vignettes/SPOTlight_kidney.Rmd" class="external-link"><code>vignettes/SPOTlight_kidney.Rmd</code></a></small>
      <div class="d-none name"><code>SPOTlight_kidney.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>
      </p>
<p>Spatially resolved gene expression profiles are key to understand tissue organization and function. However, novel spatial transcriptomics (ST) profiling techniques lack single-cell resolution and require a combination with single-cell RNA sequencing (scRNA-seq) information to deconvolute the spatially indexed datasets. Leveraging the strengths of both data types, we developed SPOTlight, a computational tool that enables the integration of ST with scRNA-seq data to infer the location of cell types and states within a complex tissue. SPOTlight is centered around a seeded non-negative matrix factorization (NMF) regression, initialized using cell-type marker genes and non-negative least squares (NNLS) to subsequently deconvolute ST capture locations (spots).</p>
    </div>
    
<style type="text/css">
.smaller {
  font-size: 10px
}
</style>
<p>For a more detailed explanation of <code>SPOTlight</code> consider looking at our manuscript: &gt; Elosua-Bayes M, Nieto P, Mereu E, Gut I, Heyn H.<br>
SPOTlight: seeded NMF regression to deconvolute<br>
spatial transcriptomics spots with single-cell transcriptomes.<br><em>Nucleic Acids Res.</em> <strong>2021;49(9):e50</strong>. doi: <a href="10.1093/nar/gkab043">10.1093</a></p>
<div class="section level2">
<h2 class="unnumbered" id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarcElosua/SPOTlight" class="external-link">SPOTlight</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SpatialExperiment</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<div class="section level3">
<h3 id="what-is-spotlight">What is <code>SPOTlight</code>?<a class="anchor" aria-label="anchor" href="#what-is-spotlight"></a>
</h3>
<p><code>SPOTlight</code> is a tool that enables the deconvolution of cell types and cell type proportions present within each capture location comprising mixtures of cells. Originally developed for 10X’s Visium - spatial transcriptomics - technology, it can be used for all technologies returning mixtures of cells.</p>
<p><code>SPOTlight</code> is based on learning topic profile signatures, by means of an NMFreg model, for each cell type and finding which combination of cell types fits best the spot we want to deconvolute. Find below a graphical abstract visually summarizing the key steps.</p>
<p><img src="../inst/extdata/schematic.png"></p>
</div>
<div class="section level3">
<h3 id="starting-point">Starting point<a class="anchor" aria-label="anchor" href="#starting-point"></a>
</h3>
<p>The minimal unit of data required to run <code>SPOTlight</code> are:</p>
<ul>
<li>ST (sparse) matrix with the expression, raw or normalized, where rows = genes and columns = capture locations.</li>
<li>Single cell (sparse) matrix with the expression, raw or normalized,<br>
where rows = genes and columns = cells.</li>
<li>Vector indicating the cell identity for each column in the single cell expression matrix.</li>
</ul>
<p>Data inputs can also be objects of class <em><a href="https://bioconductor.org/packages/3.14/SpatialExperiment" class="external-link">SpatialExperiment</a></em> (SE), <em><a href="https://bioconductor.org/packages/3.14/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em> (SCE) or <em><a href="https://CRAN.R-project.org/package=Seurat" class="external-link">Seurat</a></em> objects from which the minimal data will be extracted.</p>
</div>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<div class="section level3">
<h3 id="data-description">Data description<a class="anchor" aria-label="anchor" href="#data-description"></a>
</h3>
<p>For this vignette, we will use a SE put out by <em>10X Genomics</em> containing a Visium kidney slide. The raw data can be accessed <a href="https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Mouse_Kidney" class="external-link">here</a>.</p>
<p>SCE data comes from the <a href="https://www.nature.com/articles/s41586-020-2496-1" class="external-link"><em>The Tabula Muris Consortium</em></a> which contains &gt;350,000 cells from from male and female mice belonging to six age groups, ranging from 1 to 30 months. From this dataset we will only load the kidney subset to map it to the Visium slide.</p>
</div>
<div class="section level3">
<h3 id="loading-the-data">Loading the data<a class="anchor" aria-label="anchor" href="#loading-the-data"></a>
</h3>
<p>Both datasets are available through Biocondcutor’s <em><a href="https://bioconductor.org/packages/3.14/ExperimentHub" class="external-link">ExperimentHub</a></em> and can be loaded into R as follows:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span>

<span class="co"># initialize a Hub instance which stores a complete set of recordd</span>
<span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># retrieve any records that match our keyword(s) of interest</span>
<span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">eh</span>, <span class="st">"Tabula Muris Senis droplet Kidney"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## ExperimentHub with 7 records</span>
<span class="co">## # snapshotDate(): 2021-10-19</span>
<span class="co">## # $dataprovider: The Tabula Muris Consortium</span>
<span class="co">## # $species: Mus musculus</span>
<span class="co">## # $rdataclass: matrix, H5File, DFrame</span>
<span class="co">## # additional mcols(): taxonomyid, genome, description,</span>
<span class="co">## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,</span>
<span class="co">## #   rdatapath, sourceurl, sourcetype </span>
<span class="co">## # retrieve records with, e.g., 'object[["EH6386"]]' </span>
<span class="co">## </span>
<span class="co">##            title                                             </span>
<span class="co">##   EH6386 | Tabula Muris Senis droplet Kidney colData         </span>
<span class="co">##   EH6387 | Tabula Muris Senis droplet Kidney counts          </span>
<span class="co">##   EH6388 | Tabula Muris Senis droplet Kidney processed counts</span>
<span class="co">##   EH6389 | Tabula Muris Senis droplet Kidney rowData         </span>
<span class="co">##   EH6390 | Tabula Muris Senis droplet Kidney PCA             </span>
<span class="co">##   EH6391 | Tabula Muris Senis droplet Kidney tSNE            </span>
<span class="co">##   EH6392 | Tabula Muris Senis droplet Kidney UMAP</span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">eh</span>, <span class="st">"MouseKidneyCoronal"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## ExperimentHub with 2 records</span>
<span class="co">## # snapshotDate(): 2021-10-19</span>
<span class="co">## # $dataprovider: 10X Genomics</span>
<span class="co">## # $species: Mus musculus</span>
<span class="co">## # $rdataclass: SpatialExperiment</span>
<span class="co">## # additional mcols(): taxonomyid, genome, description,</span>
<span class="co">## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,</span>
<span class="co">## #   rdatapath, sourceurl, sourcetype </span>
<span class="co">## # retrieve records with, e.g., 'object[["EH6707"]]' </span>
<span class="co">## </span>
<span class="co">##            title                   </span>
<span class="co">##   EH6707 | MouseKidneyCoronal      </span>
<span class="co">##   EH6743 | MouseKidneyCoronal_v3.13</span></code></pre>
<p>Load the spatial data:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/helenalc/TENxVisiumData" class="external-link">TENxVisiumData</a></span><span class="op">)</span>
<span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TENxVisiumData/man/TENxVisiumData.html" class="external-link">MouseKidneyCoronal</a></span><span class="op">(</span><span class="op">)</span>
<span class="co"># Use symbols instead of Ensembl IDs as feature names</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">symbol</span></code></pre></div>
<p>Load the single cell data. Since our data comes from the <a href="https://www.nature.com/articles/s41586-020-2496-1" class="external-link">Tabula Muris Sensis</a> dataset, we can directly load the SCE object as follows:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/fmicompbio/TabulaMurisSenisData" class="external-link">TabulaMurisSenisData</a></span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TabulaMurisSenisData/man/TabulaMurisSenisDroplet.html" class="external-link">TabulaMurisSenisDroplet</a></span><span class="op">(</span>tissues <span class="op">=</span> <span class="st">"Kidney"</span><span class="op">)</span><span class="op">$</span><span class="va">Kidney</span></code></pre></div>
<p>Quick data exploration:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">free_annotation</span>, <span class="va">sce</span><span class="op">$</span><span class="va">age</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##                                                             </span>
<span class="co">##                                                                1m   3m  18m</span>
<span class="co">##   CD45                                                         11   32   76</span>
<span class="co">##   CD45    B cell                                                7    5   45</span>
<span class="co">##   CD45    NK cell                                               1    4    8</span>
<span class="co">##   CD45    T cell                                                8   18   48</span>
<span class="co">##   CD45    macrophage                                           59  132  254</span>
<span class="co">##   CD45    plasma cell                                           0    0    9</span>
<span class="co">##   Epcam     kidney distal convoluted tubule epithelial cell    78  126  169</span>
<span class="co">##   Epcam    brush cell                                          52   15   78</span>
<span class="co">##   Epcam    kidney collecting duct principal cell               77  110  132</span>
<span class="co">##   Epcam    kidney proximal convoluted tubule epithelial cell  945  684 1120</span>
<span class="co">##   Epcam    podocyte                                            92   94   64</span>
<span class="co">##   Epcam    proximal tube epithelial cell                       25  195  296</span>
<span class="co">##   Epcam    thick ascending tube S epithelial cell             465  312  248</span>
<span class="co">##   Pecam    Kidney cortex artery cell                           75   78   91</span>
<span class="co">##   Pecam    fenestrated capillary endothelial                  164  182  164</span>
<span class="co">##   Pecam    kidney capillary endothelial cell                   49   38   25</span>
<span class="co">##   Stroma    fibroblast                                         15   16   37</span>
<span class="co">##   Stroma    kidney mesangial cell                              80   51   18</span>
<span class="co">##   nan                                                         285  238  256</span>
<span class="co">##                                                             </span>
<span class="co">##                                                               21m  24m  30m</span>
<span class="co">##   CD45                                                         54 1010  106</span>
<span class="co">##   CD45    B cell                                               54 2322   62</span>
<span class="co">##   CD45    NK cell                                               2   47    4</span>
<span class="co">##   CD45    T cell                                               42  846  177</span>
<span class="co">##   CD45    macrophage                                          101  259  514</span>
<span class="co">##   CD45    plasma cell                                          12  169   42</span>
<span class="co">##   Epcam     kidney distal convoluted tubule epithelial cell   153    0  131</span>
<span class="co">##   Epcam    brush cell                                         169    0   31</span>
<span class="co">##   Epcam    kidney collecting duct principal cell               58    0  370</span>
<span class="co">##   Epcam    kidney proximal convoluted tubule epithelial cell  868    0  817</span>
<span class="co">##   Epcam    podocyte                                            66    0  170</span>
<span class="co">##   Epcam    proximal tube epithelial cell                        5    0 1977</span>
<span class="co">##   Epcam    thick ascending tube S epithelial cell             228    0  313</span>
<span class="co">##   Pecam    Kidney cortex artery cell                           69    0  115</span>
<span class="co">##   Pecam    fenestrated capillary endothelial                  134    0  211</span>
<span class="co">##   Pecam    kidney capillary endothelial cell                   18    0    7</span>
<span class="co">##   Stroma    fibroblast                                         13    0   80</span>
<span class="co">##   Stroma    kidney mesangial cell                              22    0    7</span>
<span class="co">##   nan                                                         189 1068  579</span></code></pre>
<p>We see how there is a good representation of all the cell types across ages except at 24m. In order to reduce the potential noise introduced by age and batch effects we are going to select cells all coming from the same age.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Keep cells from 18m mice</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">sce</span>, , <span class="va">age</span> <span class="op">==</span> <span class="st">"18m"</span><span class="op">)</span>
<span class="co"># Keep cells with clear cell type annotations</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">sce</span>, , <span class="op">!</span><span class="va">free_annotation</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nan"</span>, <span class="st">"CD45"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="workflow">Workflow<a class="anchor" aria-label="anchor" href="#workflow"></a>
</h2>
<div class="section level3">
<h3 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h3>
<p>If the dataset is very large we want to downsample it to train the model, both in of number of cells and number of genes. To do this, we want to keep a representative amount of cells per cluster and the most biologically relevant genes.</p>
<p>In the paper we show how downsampling the number of cells per cell identity to ~100 doesn’t affect the performance of the model. Including &gt;100 cells per cell identity provides marginal improvement while greatly increasing computational time and resources. Furthermore, restricting the gene set to the marker genes for each cell type along with up to 3.000 highly variable genes further optimizes performance and computational resources. You can find a more detailed explanation in the original <a href="https://academic.oup.com/nar/article/49/9/e50/6129341" class="external-link">paper</a>.</p>
<div class="section level4">
<h4 id="feature-selection">Feature selection<a class="anchor" aria-label="anchor" href="#feature-selection"></a>
</h4>
<p>Our first step is to get the marker genes for each cell type. We follow the Normalization procedure as described in <a href="http://bioconductor.org/books/3.14/OSCA.basic/normalization.html" class="external-link">OSCA</a>. We first carry out library size normalization to correct for cell-specific biases:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="variance-modelling">Variance modelling<a class="anchor" aria-label="anchor" href="#variance-modelling"></a>
</h4>
<p>We aim to identify highly variable genes that drive biological heterogeneity. By feeding these genes to the model we improve the resolution of the biological structure and reduce the technical noise.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">dec</span><span class="op">$</span><span class="va">mean</span>, <span class="va">dec</span><span class="op">$</span><span class="va">total</span>, xlab <span class="op">=</span> <span class="st">"Mean log-expression"</span>, ylab <span class="op">=</span> <span class="st">"Variance"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/curve.html" class="external-link">curve</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">dec</span><span class="op">)</span><span class="op">$</span><span class="fu">trend</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"blue"</span>, add <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="SPOTlight_kidney_files/figure-html/variance-1.png" width="700"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get the top 3000 genes.</span>
<span class="va">hvg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n <span class="op">=</span> <span class="fl">3000</span><span class="op">)</span></code></pre></div>
<p>Next we obtain the marker genes for each cell identity. You can use whichever method you want as long as it returns a weight indicating the importance of that gene for that cell type. Examples include <code>avgLogFC</code>, <code>AUC</code>, <code>pct.expressed</code>, <code>p-value</code>…</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-colData.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">free_annotation</span>

<span class="co"># Get vector indicating which genes are neither ribosomal or mitochondrial</span>
<span class="va">genes</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"^Rp[l|s]|Mt"</span>, x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Compute marker genes</span>
<span class="va">mgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/scoreMarkers.html" class="external-link">scoreMarkers</a></span><span class="op">(</span><span class="va">sce</span>, subset.row <span class="op">=</span> <span class="va">genes</span><span class="op">)</span></code></pre></div>
<p>Then we want to keep only those genes that are relevant for each cell identity:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mgs_fil</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mgs</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">x</span> <span class="op">&lt;-</span> <span class="va">mgs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>
    <span class="co"># Filter and keep relevant marker genes, those with AUC &gt; 0.8</span>
    <span class="va">x</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span><span class="va">x</span><span class="op">$</span><span class="va">mean.AUC</span> <span class="op">&gt;</span> <span class="fl">0.8</span>, <span class="op">]</span>
    <span class="co"># Sort the genes from highest to lowest weight</span>
    <span class="va">x</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">mean.AUC</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span>
    <span class="co"># Add gene and cluster id to the dataframe</span>
    <span class="va">x</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>
    <span class="va">x</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="va">i</span>
    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>
<span class="va">mgs_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">mgs_fil</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="cell-downsampling">Cell Downsampling<a class="anchor" aria-label="anchor" href="#cell-downsampling"></a>
</h4>
<p>Next, we randomly select at most 100 cells per cell identity. If a cell type is comprised of &lt;100 cells, all the cells will be used. If we have very biologically different cell identities (B cells vs. T cells vs. Macrophages vs. Epithelial) we can use fewer cells since their transcriptional profiles will be very different. In cases when we have more transcriptionally similar cell identities we need to increase our N to capture the biological heterogeneity between them.</p>
<p>In our experience we have found that for this step it is better to select the cells from each cell type from the same batch if you have a joint dataset from multiple runs. This will ensure that the model removes as much signal from the batch as possible and actually learns the biological signal.</p>
<p>For the purpose of this vignette and to speed up the analysis, we are going to use 20 cells per cell identity:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># split cell indices by identity</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>, <span class="va">sce</span><span class="op">$</span><span class="va">free_annotation</span><span class="op">)</span>
<span class="co"># downsample to at most 20 per identity &amp; subset</span>
<span class="va">n_cells</span> <span class="op">&lt;-</span> <span class="fl">2</span>
<span class="va">cs_keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">idx</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span>
    <span class="kw">if</span> <span class="op">(</span><span class="va">n</span> <span class="op">&lt;</span> <span class="va">n_cells</span><span class="op">)</span>
        <span class="va">n_cells</span> <span class="op">&lt;-</span> <span class="va">n</span>
    <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">i</span>, <span class="va">n_cells</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">cs_keep</span><span class="op">)</span><span class="op">]</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="deconvolution">Deconvolution<a class="anchor" aria-label="anchor" href="#deconvolution"></a>
</h3>
<p>You are now set to run <code>SPOTlight</code> to deconvolute the spots!</p>
<p>Briefly, here is how it works:</p>
<ol style="list-style-type: decimal">
<li><p>NMF is used to factorize a matrix into two lower dimensionality matrices without negative elements. We first have an initial matrix V (SCE count matrix), which is factored into W and H. Unit variance normalization by gene is performed in V and in order to standardize discretized gene expression levels, ‘counts-umi’. Factorization is then carried out using the non-smooth NMF method, implemented in the R package NMF <em><a href="https://CRAN.R-project.org/package=NMF" class="external-link">NMF</a></em> (14). This method is intended to return sparser results during the factorization in W and H, thus promoting cell-type-specific topic profile and reducing overfitting during training. Before running factorization, we initialize each topic, column, of W with the unique marker genes for each cell type with weights. In turn, each topic of H in <code>SPOTlight</code> is initialized with the corresponding belongance of each cell for each topic, 1 or 0. This way, we seed the model with prior information, thus guiding it towards a biologically relevant result. This initialization also aims at reducing variability and improving the consistency between runs.<br></p></li>
<li><p>NNLS regression is used to map each capture location’s transcriptome in V’ (SE count matrix) to H’ using W as the basis. We obtain a topic profile distribution over each capture location which we can use to determine its composition.<br></p></li>
<li><p>we obtain Q, cell-type specific topic profiles, from H. We select all cells from the same cell type and compute the median of each topic for a consensus cell-type-specific topic signature. We then use NNLS to find the weights of each cell type that best fit H’ minimizing the residuals.</p></li>
</ol>
<p>You can visualize the above explanation in the following workflow scheme:</p>
<p><img src="../inst/extdata/workflow.png"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SPOTlight.html">SPOTlight</a></span><span class="op">(</span>
    x <span class="op">=</span> <span class="va">sce</span>,
    y <span class="op">=</span> <span class="va">spe</span>,
    groups <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">free_annotation</span>,
    mgs <span class="op">=</span> <span class="va">mgs_df</span>,
    hvg <span class="op">=</span> <span class="va">hvg</span>,
    weight_id <span class="op">=</span> <span class="st">"mean.AUC"</span>,
    group_id <span class="op">=</span> <span class="st">"cluster"</span>,
    gene_id <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Scaling count matrix</span></code></pre>
<pre><code><span class="co">## Seeding initial matrices</span></code></pre>
<pre><code><span class="co">## Training NMF model</span></code></pre>
<pre><code><span class="co">## Time for training: 0.2min</span></code></pre>
<pre><code><span class="co">## Deconvoluting mixture data</span></code></pre>
<p>Extract data from <code>SPOTlight</code>:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Extract deconvolution matrix</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">mat</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">##                    CD45    B cell CD45    NK cell CD45    T cell</span>
<span class="co">## AAACCGTTCGTCCAGG-1     0.04085780      0.03745327     0.09125076</span>
<span class="co">## AAACCTAAGCAGCCGG-1     0.00000000      0.00000000     0.09927724</span>
<span class="co">## AAACGAGACGGTTGAT-1     0.05342870      0.01022877     0.11173765</span>
<span class="co">## AAACGGTTGCGAACTG-1     0.02799133      0.03798122     0.07084192</span>
<span class="co">## AAACTCGGTTCGCAAT-1     0.01718571      0.14946677     0.08317662</span>
<span class="co">## AAACTGCTGGCTCCAA-1     0.07110740      0.05131965     0.08557376</span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Extract NMF model fit</span>
<span class="va">mod</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">NMF</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<p>In the next section we show how to visualize the data and interpret <code>SPOTlight</code>’s results.</p>
<div class="section level3">
<h3 id="topic-profiles">Topic profiles<a class="anchor" aria-label="anchor" href="#topic-profiles"></a>
</h3>
<p>We first take a look at the Topic profiles. By setting <code>facet = FALSE</code> we want to evaluate how specific each topic signature is for each cell identity. Ideally each cell identity will have a unique topic profile associated to it as seen below.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotTopicProfiles.html">plotTopicProfiles</a></span><span class="op">(</span>
    x <span class="op">=</span> <span class="va">mod</span>,
    y <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">free_annotation</span>,
    facet <span class="op">=</span> <span class="cn">FALSE</span>,
    min_prop <span class="op">=</span> <span class="fl">0.01</span>,
    ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="SPOTlight_kidney_files/figure-html/plotTopicProfiles1-1.png" width="576"></p>
<p>Next we also want to ensure that all the cells from the same cell identity share a similar topic profile since this will mean that <code>SPOTlight</code> has learned a consistent signature for all the cells from the same cell identity.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotTopicProfiles.html">plotTopicProfiles</a></span><span class="op">(</span>
    x <span class="op">=</span> <span class="va">mod</span>,
    y <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">free_annotation</span>,
    facet <span class="op">=</span> <span class="cn">TRUE</span>,
    min_prop <span class="op">=</span> <span class="fl">0.01</span>,
    ncol <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></code></pre></div>
<p><img src="SPOTlight_kidney_files/figure-html/plotTopicProfiles2-1.png" width="864"></p>
<p>Lastly we can take a look at which genes the model learned for each topic. Higher values indicate that the gene is more relevant for that topic. In the below table we can see how the top genes for <code>Topic1</code> are characteristic for B cells (i.e. <em>Cd79a</em>, <em>Cd79b</em>, <em>Ms4a1</em>…).</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://renozao.github.io/NMF" class="external-link">NMF</a></span><span class="op">)</span>
<span class="va">sign</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NMF/man/basis-coef-methods.html" class="external-link">basis</a></span><span class="op">(</span><span class="va">mod</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sign</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Topic"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sign</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sign</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##            Topic1        Topic2        Topic3        Topic4      Topic5</span>
<span class="co">## Cd79a 0.002810294 4.234637e-319  0.000000e+00 6.705534e-221 0.001914301</span>
<span class="co">## Ly6d  0.004164431 2.064875e-248 2.478523e-285  0.000000e+00 0.001618620</span>
<span class="co">## Fau   0.006895937  2.490729e-03  1.553858e-03  1.942712e-03 0.001695927</span>
<span class="co">## Cd37  0.002678810 3.230637e-220  1.606590e-03  1.225580e-48 0.001991599</span>
<span class="co">## Cd79b 0.003905927 2.728267e-301  0.000000e+00 1.105812e-137 0.001339975</span>
<span class="co">## Cd74  0.001593506 1.968397e-175 2.540083e-226  1.486929e-03 0.001602849</span>
<span class="co">##              Topic6        Topic7        Topic8        Topic9       Topic10</span>
<span class="co">## Cd79a  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00</span>
<span class="co">## Ly6d  8.975658e-316  0.000000e+00 1.729897e-307 5.433815e-314  2.696091e-04</span>
<span class="co">## Fau    1.692851e-60  5.041989e-16  1.945795e-66  1.568102e-11  6.454253e-04</span>
<span class="co">## Cd37  1.404302e-287 4.579565e-150 2.706540e-279 8.501571e-286  0.000000e+00</span>
<span class="co">## Cd79b  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00</span>
<span class="co">## Cd74  8.529272e-230 7.334888e-239 1.694522e-234 1.525215e-204 6.501910e-292</span>
<span class="co">##             Topic11       Topic12       Topic13       Topic14       Topic15</span>
<span class="co">## Cd79a  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00</span>
<span class="co">## Ly6d   0.000000e+00 1.704725e-302 1.477727e-300  0.000000e+00 4.943387e-298</span>
<span class="co">## Fau    1.391518e-15  4.520660e-26  1.159283e-04  7.740333e-18  7.263718e-30</span>
<span class="co">## Cd37   0.000000e+00 2.667158e-274 2.312004e-272 3.517747e-321 7.734263e-270</span>
<span class="co">## Cd79b  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00</span>
<span class="co">## Cd74  2.976077e-301 1.747066e-243 8.821218e-161  4.980218e-20 4.804072e-225</span>
<span class="co">##             Topic16       Topic17</span>
<span class="co">## Cd79a  0.000000e+00  0.000000e+00</span>
<span class="co">## Ly6d   0.000000e+00  0.000000e+00</span>
<span class="co">## Fau    2.607292e-04  3.976858e-67</span>
<span class="co">## Cd37   0.000000e+00 9.653549e-320</span>
<span class="co">## Cd79b  0.000000e+00  0.000000e+00</span>
<span class="co">## Cd74  1.765619e-282 1.843635e-238</span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># This can be dynamically visualized with DT as shown below</span>
<span class="co"># DT::datatable(sign, fillContainer = TRUE, filter = "top")</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="spatial-correlation-matrix">Spatial Correlation Matrix<a class="anchor" aria-label="anchor" href="#spatial-correlation-matrix"></a>
</h3>
<p>See <a href="http://www.sthda.com/english/wiki/ggcorrplot-visualization-of-a-correlation-matrix-using-ggplot2" class="external-link">here</a> for additional graphical parameters.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotCorrelationMatrix.html">plotCorrelationMatrix</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></code></pre></div>
<p><img src="SPOTlight_kidney_files/figure-html/plotCorrelationMatrix-1.png" width="864"></p>
</div>
<div class="section level3">
<h3 id="co-localization">Co-localization<a class="anchor" aria-label="anchor" href="#co-localization"></a>
</h3>
<p>Now that we know which cell types are found within each spot we can make a graph representing spatial interactions where cell types will have stronger edges between them the more often we find them within the same spot.</p>
<p>See <a href="https://www.r-graph-gallery.com/network.html" class="external-link">here</a> for additional graphical parameters.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotInteractions.html">plotInteractions</a></span><span class="op">(</span><span class="va">mat</span>, <span class="st">"heatmap"</span><span class="op">)</span></code></pre></div>
<p><img src="SPOTlight_kidney_files/figure-html/plotInteractions-1.png" width="864"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotInteractions.html">plotInteractions</a></span><span class="op">(</span><span class="va">mat</span>, <span class="st">"network"</span><span class="op">)</span></code></pre></div>
<p><img src="SPOTlight_kidney_files/figure-html/plotInteractions-2.png" width="864"></p>
</div>
<div class="section level3">
<h3 id="scatterpie">Scatterpie<a class="anchor" aria-label="anchor" href="#scatterpie"></a>
</h3>
<p>We can also visualize the cell type proportions as sections of a pie chart for each spot. You can modify the colors as you would a standard <em><a href="https://CRAN.R-project.org/package=ggplot2" class="external-link">ggplot2</a></em>.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ct</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span>
<span class="va">mat</span><span class="op">[</span><span class="va">mat</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span>

<span class="co"># Define color palette</span>
<span class="co"># (here we use 'paletteMartin' from the 'colorBlindness' package)</span>
<span class="va">paletteMartin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>
    <span class="st">"#000000"</span>, <span class="st">"#004949"</span>, <span class="st">"#009292"</span>, <span class="st">"#ff6db6"</span>, <span class="st">"#ffb6db"</span>, 
    <span class="st">"#490092"</span>, <span class="st">"#006ddb"</span>, <span class="st">"#b66dff"</span>, <span class="st">"#6db6ff"</span>, <span class="st">"#b6dbff"</span>, 
    <span class="st">"#920000"</span>, <span class="st">"#924900"</span>, <span class="st">"#db6d00"</span>, <span class="st">"#24ff24"</span>, <span class="st">"#ffff6d"</span><span class="op">)</span>

<span class="va">pal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="va">paletteMartin</span><span class="op">)</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ct</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">ct</span>

<span class="fu"><a href="../reference/plotSpatialScatterpie.html">plotSpatialScatterpie</a></span><span class="op">(</span>
    x <span class="op">=</span> <span class="va">spe</span>,
    y <span class="op">=</span> <span class="va">mat</span>,
    cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span>,
    img <span class="op">=</span> <span class="cn">FALSE</span>,
    scatterpie_alpha <span class="op">=</span> <span class="fl">1</span>,
    pie_scale <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>
        values <span class="op">=</span> <span class="va">pal</span>,
        breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pal</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="SPOTlight_kidney_files/figure-html/Scatterpie-1.png" width="864"></p>
</div>
<div class="section level3">
<h3 id="residuals">Residuals<a class="anchor" aria-label="anchor" href="#residuals"></a>
</h3>
<p>Lastly we can also take a look at how well the model predicted the proportions for each spot. We do this by looking at the residuals of the sum of squares for each spot which indicates the amount of biological signal not explained by the model.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spe</span><span class="op">$</span><span class="va">res_ss</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">]</span>
<span class="va">xy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span>
<span class="va">spe</span><span class="op">$</span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">xy</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>
<span class="va">spe</span><span class="op">$</span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">xy</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/pkg/scater/man/ggsce.html" class="external-link">ggcells</a></span><span class="op">(</span><span class="va">spe</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, color <span class="op">=</span> <span class="va">res_ss</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="SPOTlight_kidney_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.2 (2021-11-01)</span>
<span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">## Running under: Ubuntu 20.04.2 LTS</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/atlas/libblas.so.3.10.3</span>
<span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/atlas/liblapack.so.3.10.3</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##  [3] LC_TIME=es_ES.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##  [5] LC_MONETARY=es_ES.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">##  [7] LC_PAPER=es_ES.UTF-8       LC_NAME=C                 </span>
<span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">## [11] LC_MEASUREMENT=es_ES.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">## [8] base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] NMF_0.23.0                  cluster_2.1.2              </span>
<span class="co">##  [3] rngtools_1.5.2              pkgmaker_0.32.2            </span>
<span class="co">##  [5] registry_0.5-1              rhdf5_2.38.0               </span>
<span class="co">##  [7] TabulaMurisSenisData_1.0.0  TENxVisiumData_1.2.0       </span>
<span class="co">##  [9] ExperimentHub_2.2.0         AnnotationHub_3.2.0        </span>
<span class="co">## [11] BiocFileCache_2.2.0         dbplyr_2.1.1               </span>
<span class="co">## [13] scran_1.22.1                scater_1.22.0              </span>
<span class="co">## [15] scuttle_1.4.0               SpatialExperiment_1.4.0    </span>
<span class="co">## [17] SingleCellExperiment_1.16.0 SummarizedExperiment_1.24.0</span>
<span class="co">## [19] GenomicRanges_1.46.1        GenomeInfoDb_1.30.0        </span>
<span class="co">## [21] IRanges_2.28.0              S4Vectors_0.32.3           </span>
<span class="co">## [23] MatrixGenerics_1.6.0        matrixStats_0.61.0         </span>
<span class="co">## [25] SPOTlight_0.99.0            Biobase_2.54.0             </span>
<span class="co">## [27] BiocGenerics_0.40.0         ggplot2_3.3.5              </span>
<span class="co">## [29] BiocStyle_2.22.0           </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] systemfonts_1.0.3             plyr_1.8.6                   </span>
<span class="co">##   [3] igraph_1.2.9                  BiocParallel_1.28.2          </span>
<span class="co">##   [5] gridBase_0.4-7                digest_0.6.29                </span>
<span class="co">##   [7] foreach_1.5.1                 htmltools_0.5.2              </span>
<span class="co">##   [9] viridis_0.6.2                 magick_2.7.3                 </span>
<span class="co">##  [11] gdata_2.18.0                  fansi_0.5.0                  </span>
<span class="co">##  [13] magrittr_2.0.1                memoise_2.0.1                </span>
<span class="co">##  [15] ScaledMatrix_1.2.0            doParallel_1.0.16            </span>
<span class="co">##  [17] limma_3.50.0                  Biostrings_2.62.0            </span>
<span class="co">##  [19] R.utils_2.11.0                pkgdown_2.0.2                </span>
<span class="co">##  [21] colorspace_2.0-2              rappdirs_0.3.3               </span>
<span class="co">##  [23] blob_1.2.2                    ggrepel_0.9.1                </span>
<span class="co">##  [25] textshaping_0.3.6             xfun_0.28                    </span>
<span class="co">##  [27] dplyr_1.0.7                   crayon_1.4.2                 </span>
<span class="co">##  [29] RCurl_1.98-1.5                jsonlite_1.7.2               </span>
<span class="co">##  [31] scatterpie_0.1.7              iterators_1.0.13             </span>
<span class="co">##  [33] glue_1.5.1                    polyclip_1.10-0              </span>
<span class="co">##  [35] gtable_0.3.0                  nnls_1.4                     </span>
<span class="co">##  [37] zlibbioc_1.40.0               XVector_0.34.0               </span>
<span class="co">##  [39] DelayedArray_0.20.0           BiocSingular_1.10.0          </span>
<span class="co">##  [41] DropletUtils_1.14.1           Rhdf5lib_1.16.0              </span>
<span class="co">##  [43] HDF5Array_1.22.1              scales_1.1.1                 </span>
<span class="co">##  [45] DBI_1.1.1                     edgeR_3.36.0                 </span>
<span class="co">##  [47] Rcpp_1.0.7                    viridisLite_0.4.0            </span>
<span class="co">##  [49] xtable_1.8-4                  dqrng_0.3.0                  </span>
<span class="co">##  [51] bit_4.0.4                     rsvd_1.0.5                   </span>
<span class="co">##  [53] httr_1.4.2                    metapod_1.2.0                </span>
<span class="co">##  [55] RColorBrewer_1.1-2            ellipsis_0.3.2               </span>
<span class="co">##  [57] farver_2.1.0                  pkgconfig_2.0.3              </span>
<span class="co">##  [59] R.methodsS3_1.8.1             sass_0.4.0                   </span>
<span class="co">##  [61] locfit_1.5-9.4                utf8_1.2.2                   </span>
<span class="co">##  [63] ggcorrplot_0.1.3              labeling_0.4.2               </span>
<span class="co">##  [65] AnnotationDbi_1.56.2          later_1.3.0                  </span>
<span class="co">##  [67] tidyselect_1.1.1              rlang_0.4.12                 </span>
<span class="co">##  [69] reshape2_1.4.4                BiocVersion_3.14.0           </span>
<span class="co">##  [71] munsell_0.5.0                 tools_4.1.2                  </span>
<span class="co">##  [73] cachem_1.0.6                  generics_0.1.1               </span>
<span class="co">##  [75] RSQLite_2.2.9                 evaluate_0.14                </span>
<span class="co">##  [77] stringr_1.4.0                 fastmap_1.1.0                </span>
<span class="co">##  [79] yaml_2.2.1                    ragg_1.2.1                   </span>
<span class="co">##  [81] knitr_1.36                    bit64_4.0.5                  </span>
<span class="co">##  [83] fs_1.5.1                      purrr_0.3.4                  </span>
<span class="co">##  [85] KEGGREST_1.34.0               sparseMatrixStats_1.6.0      </span>
<span class="co">##  [87] mime_0.12                     R.oo_1.24.0                  </span>
<span class="co">##  [89] compiler_4.1.2                png_0.1-7                    </span>
<span class="co">##  [91] interactiveDisplayBase_1.32.0 filelock_1.0.2               </span>
<span class="co">##  [93] curl_4.3.2                    beeswarm_0.4.0               </span>
<span class="co">##  [95] tweenr_1.0.2                  tibble_3.1.6                 </span>
<span class="co">##  [97] statmod_1.4.36                bslib_0.3.1                  </span>
<span class="co">##  [99] stringi_1.7.6                 highr_0.9                    </span>
<span class="co">## [101] desc_1.4.0                    lattice_0.20-45              </span>
<span class="co">## [103] bluster_1.4.0                 Matrix_1.3-4                 </span>
<span class="co">## [105] vctrs_0.3.8                   pillar_1.6.4                 </span>
<span class="co">## [107] lifecycle_1.0.1               rhdf5filters_1.6.0           </span>
<span class="co">## [109] BiocManager_1.30.16           jquerylib_0.1.4              </span>
<span class="co">## [111] BiocNeighbors_1.12.0          bitops_1.0-7                 </span>
<span class="co">## [113] irlba_2.3.4                   httpuv_1.6.3                 </span>
<span class="co">## [115] R6_2.5.1                      promises_1.2.0.1             </span>
<span class="co">## [117] bookdown_0.24                 gridExtra_2.3                </span>
<span class="co">## [119] vipor_0.4.5                   codetools_0.2-18             </span>
<span class="co">## [121] MASS_7.3-54                   gtools_3.9.2                 </span>
<span class="co">## [123] assertthat_0.2.1              rprojroot_2.0.2              </span>
<span class="co">## [125] rjson_0.2.20                  withr_2.4.3                  </span>
<span class="co">## [127] SeuratObject_4.0.4            GenomeInfoDbData_1.2.7       </span>
<span class="co">## [129] parallel_4.1.2                ggfun_0.0.4                  </span>
<span class="co">## [131] grid_4.1.2                    beachmat_2.10.0              </span>
<span class="co">## [133] tidyr_1.1.4                   rmarkdown_2.11               </span>
<span class="co">## [135] DelayedMatrixStats_1.16.0     ggforce_0.3.3                </span>
<span class="co">## [137] shiny_1.7.1                   ggbeeswarm_0.6.0</span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Marc Elosua-Bayes, Helena L. Crowell.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
